apps/frontend/src/utils/ZwoParsing.ts

```diff
@@ -0,0 +1,235 @@
+import { parseXml, XmlDocument, XmlElement, XmlText } from '@rgrove/parse-xml';
+import { Workout, WorkoutPart } from '../types';
+
+export const parseZwoWorkout = (
+  str: string
+): { workout: Workout | null; errors: Set<string> } => {
+  try {
+    const document: XmlDocument = parseXml(str);
+    const workoutFileElement = document.children[0];
+    if (!isXmlElement(workoutFileElement)) {
+      return {
+        workout: null,
+        errors: new Set([`Invalid .zwo file: Missing children element`]),
+      };
+    }
+    const zwoWorkout = parseWorkoutFile(workoutFileElement);
+    return {
+      workout: zwoWorkout.workout
+        ? zwoWorkoutToDundringWorkout(zwoWorkout.workout)
+        : null,
+      errors: zwoWorkout.errors,
+    };
+  } catch (err) {
+    console.error(err);
+    return {
+      workout: null,
+      errors: new Set([`Invalid .zwo file: ${err}`]),
+    };
+  }
+};
+
+type ZwoWorkout = {
+  name: string;
+  parts: Array<ZwoWorkoutPart>;
+};
+
+type ZwoWorkoutPart = SteadyState | IntervalsT | FreeRide | Warmup | Cooldown;
+
+type SteadyState = { type: 'steadystate'; duration: number; power: number };
+
+type IntervalsT = {
+  type: 'intervals';
+  repeat: number;
+  onDuration: number;
+  offDuration: number;
+  onPower: number;
+  offPower: number;
+};
+
+type FreeRide = { type: 'freeride'; duration: number };
+
+type Warmup = {
+  type: 'warmup' | 'ramp';
+  duration: number;
+  powerLow: number;
+  powerHigh: number;
+};
+
+type Cooldown = {
+  type: 'cooldown';
+  duration: number;
+  powerLow: number;
+  powerHigh: number;
+};
+
+const parseWorkoutFile = (
+  workoutFileElement: XmlElement
+): { workout: ZwoWorkout | null; errors: Set<string> } => {
+  const workoutElement = findXmlElementChildWithName(
+    workoutFileElement,
+    'workout'
+  );
+  if (!workoutElement) {
+    return {
+      workout: null,
+      errors: new Set([`Invalid ZWO-file: ${'workout-element is missing'}`]),
```

Comment:
Why is this a template string?
```suggestion
      errors: new Set(['Invalid ZWO-file: workout-element is missing']),
```